crates/autopilot/src/solvable_orders.rs

```diff
@@ -506,12 +513,32 @@ async fn find_invalid_signature_orders(
 /// Removes orders that can't possibly be settled because there isn't enough
 /// balance.
 fn orders_with_balance(mut orders: Vec<Order>, balances: &Balances) -> Vec<Order> {
-    // Prefer newer orders over older ones.
-    orders.sort_by_key(|order| std::cmp::Reverse(order.metadata.creation_date));
     orders.retain(|order| {
         let balance = match balances.get(&Query::from_order(order)) {
             None => return false,
-            Some(balance) => *balance,
+            Some(balance) => balance.balance,
+        };
+
+        if order.data.partially_fillable && balance >= 1.into() {
+            return true;
+        }
+
+        let needed_balance = match order.data.sell_amount.checked_add(order.data.fee_amount) {
+            None => return false,
+            Some(balance) => balance,
+        };
+        balance >= needed_balance
+    });
+    orders
+}
+
+/// Removes orders that can't possibly be settled because the allowance is not
+/// enough
+fn orders_with_allowance(mut orders: Vec<Order>, balances: &Balances) -> Vec<Order> {
```

Comment:
These functions are on the hot path and are now iterating over the majority of orders twice and doing a hashmap lookup for each one when it could happen in one sweep.
This also results in an almost exactly duplicated function.
Did you already check how you could avoid that?

I believe you could pass in the `OrderFilterCounter` and use `checkpoint_by_invalid_orders()` inside by collecting missing balances uids and missing allowance uids into 2 separate vectors.